Animate Link ID with patient sheet

vinaysshenoy opened the pull request:
Pivotal Story Link: https://www.pivotaltracker.com/story/show/165489494

sanchita commented on app/src/main/java/org/simple/clinic/summary/linkId/LinkIdWithPatientSheetController.kt:
```diff
@@ -25,7 +25,16 @@ class LinkIdWithPatientSheetController @Inject constructor(
+        .ofType<LinkIdWithPatientSheetCreated>()
+        .map { it.identifier }
+        .map { identifier -> { ui: Ui -> ui.renderIdentifierText(identifier) } }
   }
```
I think that the controller is not doing anything here except rendering the UI from an event which is sent from the view itself.
We could avoid all this and call the `renderIdentifierText` method from `onFinishInflate` of `LinkIdWithPatientView`.

vinaysshenoy replied:
True, but I added this so that the `LinkIdWithPatientShown` event is sent to Heap for analytics. Since this is no longer an activity or a screen, the only way to send this event to Heap is via the controller.

sanchita commented on app/src/main/java/org/simple/clinic/summary/linkId/LinkIdWithPatientSheetController.kt:
```diff
@@ -10,7 +10,7 @@ import org.simple.clinic.patient.PatientRepository
+typealias Ui = LinkIdWithPatientView
 typealias UiChange = (Ui) -> Unit
 
 class LinkIdWithPatientSheetController @Inject constructor(
```
Please rename the class`LinkIdWithPatientSheetController` to `LinkIdWithPatientViewController`

vinaysshenoy replied:
Done.

sanchita commented on app/src/main/java/org/simple/clinic/summary/linkId/LinkIdWithPatientView.kt:
```diff
@@ -0,0 +1,118 @@
+        .subscribe { uiChange -> uiChange(this) }
+  }
+
+  private fun viewCreates(): Observable<UiEvent> {
```
Please rename `viewCreates` to `viewShows`

sanchita commented on app/src/test/java/org/simple/clinic/summary/PatientSummaryScreenControllerTest.kt:
```diff
@@ -1031,8 +1031,20 @@ class PatientSummaryScreenControllerTest {
+
+    uiEvents.onNext(PatientSummaryLinkIdCompleted)
+
+    verify(screen).hideLinkIdWithPatientSheet()
```
Can you also add another verify ->
`verify(screen, never()).showLinkIdWithPatientSheet(any(), any())`

vinaysshenoy replied:
That verification will always fail because the sheet must always be shown when the intention is `LinkIdWithPatient`.

sanchita commented:
The review comments left above are for PR -> https://github.com/simpledotorg/simple-android/pull/609.
I added them in this PR by mistake.
Please make changes there and resolve it here.
Sorry for the inconvenience :/

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/linkId/LinkIdWithPatientEvents.kt:
```diff
@@ -5,12 +5,23 @@ import org.simple.clinic.widgets.UiEvent
+  override val analyticsName = "LinkIdWithPatient:Cancel Clicked"
+}
+
+data class LinkIdWithPatientSheetClosed(val actionTaken: ActionTaken) : UiEvent {
```
Makes sense. I'll do this.

vinaysshenoy commented on app/src/main/AndroidManifest.xml:
```diff
@@ -72,11 +72,6 @@
       android:theme="@style/Clinic.V2.Theme.BottomSheetActivity"
```
I'll move this explanation to the KDoc for `LinkIdWithPatientView`.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:
```diff
@@ -203,11 +205,21 @@ class PatientSummaryScreen(context: Context, attrs: AttributeSet) : RelativeLayo
+    return linkIdWithPatientView
+        .uiEvents()
+        .ofType<LinkIdWithPatientSheetClosed>()
+        .filter { it.actionTaken == ActionTaken.LINKED }
```
Yup. It's been done.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:
```diff
@@ -370,9 +382,13 @@ class PatientSummaryScreen(context: Context, attrs: AttributeSet) : RelativeLayo
-    activity.startActivityForResult(intent, REQCODE_LINK_ID)
+  fun showLinkIdWithPatientSheet(patientUuid: UUID, identifier: Identifier) {
+    linkIdWithPatientView.downstreamUiEvents.onNext(LinkIdWithPatientSheetCreated(patientUuid, identifier))
+    linkIdWithPatientView.visibility = VISIBLE
```
Done.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:
```diff
@@ -370,9 +382,13 @@ class PatientSummaryScreen(context: Context, attrs: AttributeSet) : RelativeLayo
-  fun openLinkIdWithPatientSheet(patientUuid: UUID, identifier: Identifier) {
-    val intent = LinkIdWithPatientSheet.intent(context, patientUuid, identifier)
-    activity.startActivityForResult(intent, REQCODE_LINK_ID)
+  fun showLinkIdWithPatientSheet(patientUuid: UUID, identifier: Identifier) {
```
Done.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:
```diff
@@ -370,9 +382,13 @@ class PatientSummaryScreen(context: Context, attrs: AttributeSet) : RelativeLayo
+    linkIdWithPatientView.visibility = VISIBLE
+  }
+
+  fun hideLinkIdWithPatientSheet() {
```
Done.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt:
```diff
@@ -87,7 +87,8 @@ class PatientSummaryScreenController @Inject constructor(
         goBackWhenBackClicked(replayedEvents),
         goToHomeOnDoneClick(replayedEvents),
-        goBackIfLinkIdWithPatientWasCancelled(replayedEvents)
+        goBackIfLinkIdWithPatientWasCancelled(replayedEvents),
```
Done.

vinaysshenoy commented:
> The review comments left above are for PR -> #609.
> I added them in this PR by mistake.
> Please make changes there and resolve it here.
> Sorry for the inconvenience :/

@sanchita Addressed feedback in #609

sanchita commented on app/src/main/java/org/simple/clinic/widgets/BottomSheetAnimations.kt:
```diff
@@ -0,0 +1,56 @@
+ * LinkIdWithPatientView, which is not a BottomSheetActivity, but needs to behave like one.
+ **/
+
+fun showBottomSheet(
```
We should rename this method to specify that we are adding animations using this method.
Wdyt about `showBottomSheetWithAnimation`?

sanchita commented on app/src/main/java/org/simple/clinic/widgets/BottomSheetAnimations.kt:
```diff
@@ -0,0 +1,56 @@
+  }
+}
+
+fun hideBottomSheet(
```
Rename this to `hideBottomSheetWithAnimation` ?
